### PR TITLE
CI: split x86_64-mingw job

### PR DESCRIPTION
**Pull request by marcoieni**
Split the x86_64-mingw job in two to move it to free runners and reduce the use of large runners in CI.

try-job: x86_64-mingw-1
try-job: x86_64-mingw-2

### CODE REVIEW

**Comment by mati865** [comment]
`NO_DOWNLOAD_CI_LLVM: 1` can be removed, I forgot about it when upgrading the toolchain.

**Comment by MarcoIeni** [comment]
thanks! I will create a separate PR for this.
This environment variable is present in the following windows jobs:
- i686-mingw
- x86_64-mingw
- dist-i686-mingw
- dist-x86_64-mingw

Can I delete it for all of them?

**Comment by mati865** [comment]
Yeah, they all should be fine to remove.

**Comment by MarcoIeni** [comment]
Interestingly I found https://github.com/rust-lang/rust/pull/112645 which does the opposite of this PR. My goal is to try to "revert" part of the PR by using two free runners.
I'll try bors again because I don't get why it failed. Maybe it's a flaky error.

**Comment by MarcoIeni** [comment]
I'll give this one last try 🤔

**Comment by mati865** [comment]
This not a spirituous failure, the linker is misconfigured.

**Comment by MarcoIeni** [comment]
> This not a spurious failure, the linker is misconfigured.


ok, thank you. Is there something I'm missing? Because I don't understand what I did wrong

**Comment by MarcoIeni** [comment]
Maybe `make ci-mingw-x` does some work required for `make ci-mingw-bootstrap`, too?

**Comment by mati865** [comment]
I'm afraid I don't remember how all this works (especially after the changes this year).
In a working build during LLVM build you can see:
```
  -- The C compiler identification is GNU 14.1.0
  -- The CXX compiler identification is GNU 14.1.0
  -- The ASM compiler identification is GNU
  -- Found assembler: C:/a/rust/rust/mingw64/bin/gcc.exe
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: C:/a/rust/rust/mingw64/bin/gcc.exe - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: C:/a/rust/rust/mingw64/bin/g++.exe - skipped
```
Here it shows:
```
  -- The C compiler identification is Clang 18.1.4
  -- The CXX compiler identification is Clang 18.1.4
  -- The ASM compiler identification is Clang with MSVC-like command-line
  -- Found assembler: C:/a/rust/rust/citools/clang-rust/bin/clang-cl.exe
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: C:/a/rust/rust/citools/clang-rust/bin/clang-cl.exe - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: C:/a/rust/rust/citools/clang-rust/bin/clang-cl.exe - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
```
Which is almost as wrong as it can be.

**Comment by MarcoIeni** [comment]
I'm comparing the logs of a successful auto build and this try build. The linker is indeed set differently
image

**Comment by MarcoIeni** [comment]
Maybe that's the issue:

https://github.com/rust-lang/rust/blob/251dc8ad84492c792a7600d8c5fef2ec868a36a7/src/ci/shared.sh#L55-L57

File: src/ci/shared.sh [hunk 1]
```diff
@@ -53,7 +53,8 @@ function isLinux {
 }
 
 function isKnownToBeMingwBuild {
-    isGitHubActions && [[ "${CI_JOB_NAME}" == *mingw ]]
+    # CI_JOB_NAME must contain "mingw" to be considered a MinGW build.
+    isGitHubActions && [[ "${CI_JOB_NAME}" == *mingw* ]]
```
Comment (MarcoIeni) [inline src/ci/shared.sh:57]: This is going to affect the job `mingw-check`, too. Is this fine?

https://github.com/rust-lang/rust/blob/251dc8ad84492c792a7600d8c5fef2ec868a36a7/src/ci/github-actions/jobs.yml#L221

**Comment by MarcoIeni** [comment]
I ran `bors try` when another "try" was already running because I wanted to check what would happen 👀

File: src/bootstrap/mk/Makefile.in [hunk 2]
```diff
@@ -108,11 +108,16 @@ ci-msvc: ci-msvc-py ci-msvc-ps1
...
+TESTS_IN_MINGW_2 := \
+	tests/ui
+
 # test both x and bootstrap entrypoints
 ci-mingw-x:
-	$(Q)$(CFG_SRC_DIR)/x test --stage 2 tidy
+	$(Q)$(CFG_SRC_DIR)/x test --stage 2 $(TESTS_IN_2:%=--exclude %)
 ci-mingw-bootstrap:
-	$(Q)$(BOOTSTRAP) test --stage 2 --skip tidy
+	$(Q)$(BOOTSTRAP) test --stage 2 $(TESTS_IN_2)
```
Comment (MarcoIeni) [inline src/bootstrap/mk/Makefile.in:120]: This is taken from https://github.com/rust-lang/rust/pull/112633/files

**Comment by MarcoIeni** [comment]
No improvements at all in timings. I wonder why 🤔
image

**Comment by MarcoIeni** [comment]
oh, there was a typo in my PR 🙈 fixing it and trying again

**Comment by MarcoIeni** [comment]
image
Dividing the tests worked, but I need to balance them better

**Comment by MarcoIeni** [comment]
In the meantime, I want to try how long it would take for mingw-2 to run on the free runner

**Comment by Kobzol** [comment]
> Dividing the tests worked, but I need to balance them better

I think that this division is already good enough, tbh.

File: src/ci/shared.sh (hunk 1 above)
Comment (Kobzol) [reply src/ci/shared.sh:57]: It does seem to affect some nontrivial behavior, so I would rather keep the original behavior for `mingw-check` in this PR. We can just hard-code it here.

**Comment by MarcoIeni** [comment]
> I think that this division is already good enough, tbh.

One job took 1h 4m and the other 1h 48m. I was hoping to find a better balance given that the goal is to move to free runners.

Btw the try build took 3 hours to finish on a free runner. 
image
This means the only way to use free runners here is probably doing fanout, i.e. building the compiler on a large runner and running tests on free ones.

**Comment by MarcoIeni** [comment]
Actually I had a look at the logs and I saw that llvm took 96 minutes to compile 😵 Maybe sccache didn't work because of the change to the free image? I'm giving this another try.

File: src/ci/shared.sh (hunk 1 above)
Comment (MarcoIeni) [reply src/ci/shared.sh:57]: we could do:

- [ ] match on the trailing number (i.e. `-N`)

Suggested by JD 🙌

Comment (Kobzol) [reply src/ci/shared.sh:57]: Why not :) Anything works.

**Comment by MarcoIeni** [comment]
That's very encouraging! Maybe we can split this job in two free runners without fanout (at least initially to give us a quick win).
image
In the latest 3 commits I pushed, I try to balance tests better across the two jobs and tried to enable llvm downloads.

**Comment by MarcoIeni** [comment]
I will cancel this try job because I forgot to also test `x86_64-mingw-1`

**Comment by MarcoIeni** [comment]
I wonder is this is a flaky error

**Comment by Kobzol** [comment]
The error didn't look spurious, more like some Makefile definition issue.

**Comment by MarcoIeni** [comment]
ohh I get the error now, I'll fix it

File: src/ci/shared.sh (hunk 1 above)
Comment (MarcoIeni) [reply src/ci/shared.sh]: done in https://github.com/rust-lang/rust/pull/132721/commits/0f176c85ccce7c64a5be1cc0990f574d9056cde5 👍

**Comment by MarcoIeni** [comment]
Apparently downloading llvm doesn't work. I reverted the commit.

**Comment by MarcoIeni** [comment]
mingw-1 took a long time because it took 52 minutes to build llvm. Now there should be a cache it. I'll try bors again 👍

**Comment by MarcoIeni** [comment]
In the last build, mingw-1 took almost 3 hours

image

LLVM took 15 minutes to build, so I don't think that's the issue. We could try to split better the tests.

**Comment by MarcoIeni** [comment]
I confirm the test split mechanism is not working as expected.

E.g. the test `src/tools/rust-analyzer` is not executed anywhere, while the test `src/tools/cargo` runs in both runners.

Are there differences between bootstrap and x?

**Comment by Kobzol** [comment]
`x` is just a small wrapper that executes bootstrap. The problem is most likely with bootstrap itself, the code/tests that get executed based on `x test <foo>` is not always completely intuitive, and it does a bunch of implicit stuff. So it would need to be investigated on the bootstrap side.

**Comment by MarcoIeni** [comment]
Seems a flaky error.

**Comment by MarcoIeni** [comment]
image
Not bad at all!
I can balance the tests better to arrive at ~ 2h 25m, which is lower than ~ 2h 30m of the average auto build (based on 30 auto builds analyzed last week).
Before, this job was taking 2 hours on average.
Jakub, what do you think about this approach and PR overall? I know we are near the limit of what we can do. 😅

File: src/bootstrap/mk/Makefile.in [hunk 3]
```diff
@@ -108,11 +108,21 @@ ci-msvc: ci-msvc-py ci-msvc-ps1
...
+	tests/ui-fulldeps \
+	tests/rustdoc-json
+
 # test both x and bootstrap entrypoints
 ci-mingw-x:
-	$(Q)$(CFG_SRC_DIR)/x test --stage 2 tidy
+	$(Q)$(CFG_SRC_DIR)/x test --stage 2 $(TESTS_IN_MINGW_2)
 ci-mingw-bootstrap:
-	$(Q)$(BOOTSTRAP) test --stage 2 --skip tidy
+	$(Q)$(BOOTSTRAP) test --stage 2 $(TESTS_IN_MINGW_2:%=--skip %)
```
Comment (Kobzol) [inline src/bootstrap/mk/Makefile.in:125]: What does this magic do? Prefix `--skip` before each line? We should probably wait for https://github.com/rust-lang/rust/pull/132979 with this, btw.

Comment (MarcoIeni) [reply src/bootstrap/mk/Makefile.in:125]: Yeah, makes sense to wait. The best would be if they support regex. It would be nice to do `tests/*`.

**Comment by MarcoIeni** [comment]
Asked for help in [zulip](https://rust-lang.zulipchat.com/#narrow/channel/326414-t-infra.2Fbootstrap/topic/Splitting.20CI.20load.20across.20multiple.20jobs/near/483518656) to check if `tests/*` is supported

**Comment by MarcoIeni** [comment]
It didn't work. Maybe it's not possible to do `--skip tests`.
image

File: src/bootstrap/src/core/builder/mod.rs [hunk 4]
```diff
@@ -258,11 +258,11 @@ impl PathSet {
...
-        if let Some(p_kind) = &p.kind {
-            p.path.ends_with(needle) && *p_kind == module
-        } else {
-            p.path.ends_with(needle)
-        }
+        let check_path = || {
+            // This order is important for retro-compatibility, as `starts_with` was introduced later.
+            p.path.ends_with(needle) || p.path.starts_with(needle)
+        };
+        if let Some(p_kind) = &p.kind { check_path() && *p_kind == module } else { check_path() }
```
Comment (MarcoIeni) [inline src/bootstrap/src/core/builder/mod.rs:265]: this has a separate pr. I put it here just for testing

**Comment by MarcoIeni** [comment]
promising!
image

**Comment by MarcoIeni** [comment]
Yay, this looks good
image

**Comment by MarcoIeni** [comment]
force pushed, updated with master. Now I'm running bors try for the last time to check if everything is still ok 👍

**Pull request merged**